### ISSUE TITLE
sample: output converges after 2 tries

### ISSUE THREAD

**Issue by zimbatm** [nix-community/nixpkgs-fmt#132]
This sample takes two tries to stabilize

## Input
```nix
{
  buildInputs = [
  ] ++ stdenv.lib.optionals enableGui (with qt5; [ qtbase qtwebkit ])
    ++ stdenv.lib.optionals enableJupyter [ boost jsoncpp openssl zmqpp ]
  ;
}
```

## Output
```nix
{
  buildInputs = []
  ++ stdenv.lib.optionals enableGui (with qt5; [ qtbase qtwebkit ])
  ++ stdenv.lib.optionals enableJupyter [ boost jsoncpp openssl zmqpp ]
  ;
}

```

After feeding the output back as an input:

```nix
{
  buildInputs = []
    ++ stdenv.lib.optionals enableGui (with qt5; [ qtbase qtwebkit ])
    ++ stdenv.lib.optionals enableJupyter [ boost jsoncpp openssl zmqpp ]
    ;
}
```

**Comment by matklad** [comment]
Heh, that's nasty...

So, what happens is that we have special rule for `++` which indents *conditionally* based on the first element

```
ok1 = 
  []
  ++ [] # indent
  ; 

ok2 = [
]
++ [] # no indent
```

But removing newline in `[\n]` flips us from the second case to the first, and we don't detect it.  I wonder if this would be easiest to fix by introducing a separate formatting tree, a-la IntelliJ, which would allow us to make these two rules dependent on each over.

**Comment by matklad** [comment]
Hm, actually this should be not that hard to fix...

**Comment by zimbatm** [comment]
Or maybe it's a sign to drop the `++` rule. I quite like it but it seems to be causing most of the submitted issues.

**Comment by matklad** [comment]
Yeah, I would prefer that we drop that rule.  Another reason to do that is that is to minimize the diff. If you refcator nix code and change `= [` to `= \n`, the current rule shifts the indentation of all the children as well.

I've also implemneted the fix for the current rule, it wasn't as bad as I feared: https://github.com/matklad/nix-fmt/commit/2fe0e5d17bf0538e341023e69069798e6fbf6478. But I still think we should just context-independently indent the stuff
